sumpy/expansion/__init__.py

```diff
@@ -304,135 +378,159 @@ def get_coefficient_matrix(self, rscale):
+            return any(mi[d] < max_mi[d] for d in range(self.dim))
+
+        stored_identifiers = [mi for mi in mis if is_stored(mi)]
+        stored_ident_enumerate_dict = dict((tuple(mi), i) for
+                                            (i, mi) in enumerate(stored_identifiers))
+
+        coeff_matrix_dict = defaultdict(lambda: defaultdict(lambda: 0))
+        reconstruct_matrix = []
+        for i, mi in enumerate(mis):
+            reconstruct_matrix.append([])
+            if is_stored(mi):
+                coeff_matrix_dict[i][stored_ident_enumerate_dict[mi]] = 1
+                continue
+            diff = [mi[d] - max_mi[d] for d in range(self.dim)]
+            for other_mi, coeff in iteritems(pde_dict):
+                j = coeff_ident_enumerate_dict[add_mi(other_mi, diff)]
+                if i == j:
+                    continue
+                # PDE might not have max_mi_coeff = -1, divide by -max_mi_coeff
+                # to get a relation of the form, u_zz = - u_xx - u_yy for Laplace 3D.
+                reconstruct_matrix[i].append((j, coeff*max_mi_mult))
+                # j can be a stored or a non-stored multi-index
+                # Look at the coeff_matrix of j to get the j as a linear combination
+                # of stored coefficients.
+                for dep, other_coeff in iteritems(coeff_matrix_dict[j]):
+                    coeff_matrix_dict[i][dep] += other_coeff*coeff*max_mi_mult
+
+        # coeff_matrix is a dictionary of lists. Each key in the dictionary
+        # is a row number and the values are pairs of column number and value.
         coeff_matrix = defaultdict(list)
-        for i, row in iteritems(coeff_matrix_transpose):
-            for j, val in row:
-                coeff_matrix[j].append((i, val))
+        for row, deps in iteritems(coeff_matrix_dict):
+            for col, val in iteritems(deps):
+                if val != 0:
+                    coeff_matrix[row].append((col, val))
 
-        logger.debug("computing recurrence for Taylor coefficients: "
-                     "done after {dur:.2f} seconds"
-                     .format(dur=time.time() - start_time))
+        plog.done()
 
-        logger.debug("number of Taylor coefficients was reduced from {orig} to {red}"
+        print("number of Taylor coefficients was reduced from {orig} to {red}"
                      .format(orig=len(self.get_full_coefficient_identifiers()),
                              red=len(stored_identifiers)))
 
-        return stored_identifiers, coeff_matrix
+        return stored_identifiers, coeff_matrix, reconstruct_matrix
 
-    def try_get_recurrence_for_derivative(self, deriv, in_terms_of):
-        """
-        :arg deriv: a tuple of integers identifying a derivative for which
-            a recurrence is sought
-        :arg in_terms_of: a container supporting efficient containment testing
-            indicating availability of derivatives for use in recurrences
+    def get_pde(self):
```

Comment:
I think `get_pde` would do well living near the top of all this, to match flow of information match reading order.